### ISSUE TITLE
Poor performance when we use mx.symbol.Variable(..,..,stype="row-sparse") to define the variable

### ISSUE THREAD

**Issue by huifengguo** [apache/incubator-mxnet#17728]
## Description
Task: CTR prediction
Model: DCN
Dataset: Criteo
Everything is same except the definition of embedding variable:
```
// default
embed_weight = mx.symbol.Variable('embed_weight', stype='default')
// row_sparse
embed_weight = mx.symbol.Variable('embed_weight', stype='row_sparse')

```
`
embed = mx.symbol.Embedding(data=ids, weight=embed_weight,
                                               input_dim=200000, output_dim=hidden_units[0], sparse_grad=True)
`
## Occurrences
When stype="default", the result of AUC is 0.802,
while the test result of AUC is 0.794 when stype="row-sparse".

## What have you tried to solve it?

1.
2.

**Comment by eric-haibin-lin** [comment]
what optimizer are you using? update rules for some optimizer with sparse grad maybe different: https://mxnet.apache.org/api/python/docs/api/optimizer/index.html#mxnet.optimizer.SGD

**Comment by huifengguo** [comment]
> what optimizer are you using? update rules for some optimizer with sparse grad maybe different: https://mxnet.apache.org/api/python/docs/api/optimizer/index.html#mxnet.optimizer.SGD

The optimizer we are using is Adam, which is defined as:
`optim = mx.optimizer.create(optimizer, learning_rate=lr,wd=1.0E-9, rescale_grad=1.0/batch_size,epsilon=5e-8)`

**Comment by eric-haibin-lin** [comment]
you can set lazy_update=False for Adam to force disable sparse update. This will give you the same result as dense weight/gradient

**Comment by huifengguo** [comment]
> you can set lazy_update=False for Adam to force disable sparse update. This will give you the same result as dense weight/gradient

Do you mean set lazy_update=False for Adam is equivalent to forcing disable sparse update?  I always think above two setting are indenpent with each other.

**Comment by eric-haibin-lin** [comment]
Yes, if you look at the Adam optimization algorithm, zero grad does not lead to zero update. But SGD and Adagrad does.

**Comment by huifengguo** [comment]
> Yes, if you look at the Adam optimization algorithm, zero grad does not lead to zero update. But SGD and Adagrad does.

So we have to live with the loss of accuracy if we use Adam optimization algorithm and row-sparse variable?

**Comment by eric-haibin-lin** [comment]
Yes, there's no free food

**Comment by huifengguo** [comment]
> Yes, there's no free food. You can still "use Adam and row_sparse variable" with `lazy_update=False` to get identical update, the only benefit you get is less that data is transmitted during communication

OK，we just notice another setting in optimizer, namely whether we set the kvstore when initializing the optimizer.
```
optim = mx.optimizer.create(optimizer, learning_rate=lr, rescale_grad=1.0/batch_size, lazy_update=False)
mod.init_optimizer(kvstore=store, optimizer=optim)
```
Then, to verity why the loss of accuracy comes from, we conduct the following experimets. There are 3 parameters we can set, so there are 8 different settings. 

![image](https://user-images.githubusercontent.com/20381178/76191822-328d9500-621b-11ea-965c-39dc932335bd.png)

As the results show, the performance is bad when we choose kvstore and adopt row_sparse variable. Is there anything wrong? Waiting for your reply.

**Comment by huifengguo** [comment]
> Yes, there's no free food. You can still "use Adam and row_sparse variable" with `lazy_update=False` to get identical update, the only benefit you get is less that data is transmitted during communication
OK, we just notice another parameter of kvstore in mx.init_optimizer(). 
No matter we set Lazy update as True or False, when we set mod.init_optimizer as following and variable as row_sparse, the auc is only 0.794 while the auc is 0.804+ in other cases.
```
mx.mod.Module(....)
store = kv.create('local_allreduce_device')
optim = mx.optimizer.create(optimizer, learning_rate=lr,wd=1.0E-9, rescale_grad=1.0/batch_size,epsilon=5e-8, lazy_update=False)
mod.init_optimizer(kvstore=store, optimizer=optim)
```

**Comment by eric-haibin-lin** [comment]
did you pull all the params when you do evaluation?

**Comment by huifengguo** [comment]
> did you pull all the params when you do evaluation?

We conduct evaluation with the following code: 
`score = mod.score(eval_iter, ['auc','log_loss'],reset=False)`
Should we pull all the params when we run the code in single card and , BTW, how to pull all the params?

**Comment by eric-haibin-lin** [comment]
https://github.com/apache/incubator-mxnet/blob/master/example/sparse/factorization_machine/train.py#L140

**Comment by huifengguo** [comment]
> https://github.com/apache/incubator-mxnet/blob/master/example/sparse/factorization_machine/train.py#L140

I see, many thanks!
